### PR TITLE
Proposal: `fold_self` and `try_fold_self` for Iterators

### PR DESCRIPTION
**Pull request by Lucretiel**
This pull request proposes & implements two new methods on Iterators: `fold_self` and `try_fold_self`. These are variants of `fold` and `try_fold` that use the first element in the iterator as the initial accumulator.

Let me know if a public feature like this requires an RFC, or if this pull request is sufficient as place for discussion.

### CODE REVIEW

File: src/libcore/iter/iterator.rs [hunk 1]
```diff
@@ -1586,6 +1586,46 @@ pub trait Iterator {
...
+    /// assert_eq!(find_max(a.iter()), Some(Some(20)));
+    /// assert_eq!(find_max(b.iter()), Some(None));
+    /// assert_eq!(find_max(c.iter()), None);
+    /// ```
+    #[inline]
+    #[unstable(feature = "iterator_fold_self", issue = "60103")]
+    fn try_fold_self<F, R>(&mut self, mut f: F) -> Option<R>
+        where Self: Sized,
+              F: FnMut(Self::Item, Self::Item) -> R,
+              R: Try<Ok=Self::Item>
```
Comment (Centril) [inline src/libcore/iter/iterator.rs]:
```suggestion
              R: Try<Ok = Self::Item>
```

File: src/libcore/iter/iterator.rs [hunk 2]
```diff
@@ -1696,6 +1736,41 @@ pub trait Iterator {
...
+    /// iterator as the initial value, folding every subsequent element into it.
+    /// If the iterator is empty, return `None`; otherwise, return the result
+    /// of the fold.
+    ///
+    /// # Example
+    ///
+    /// Find the maximum value:
+    ///
+    /// ```
+    /// fn find_max<I>(iter: I) -> Option<I::Item>
```
Comment (Centril) [inline src/libcore/iter/iterator.rs]: Aside from the behavior for `Equal`, this is just `Iterator::max` -- maybe a different example can be provided?

File: src/libcore/iter/iterator.rs [hunk 3]
```diff
@@ -1696,6 +1736,41 @@ pub trait Iterator {
...
+    /// }
+    /// let a = [10, 20, 5, -23, 0];
+    /// let b = [];
+    ///
+    /// assert_eq!(find_max(a.iter()), Some(20));
+    /// assert_eq!(find_max(b.iter()), None));
+    /// ```
+    #[inline]
+    #[unstable(feature = "iterator_fold_self", issue = "60103")]
+    fn fold_self<F>(mut self, mut f: F) -> Option<Self::Item>
```
Comment (Centril) [inline src/libcore/iter/iterator.rs]: For reference, in itertools, this is https://docs.rs/itertools/0.8.0/itertools/trait.Itertools.html#method.fold1 which is inspired by https://hackage.haskell.org/package/base-4.12.0.0/docs/Prelude.html#v:foldl1 in Haskell.

**Comment by Dylan-DPC** [comment]
ping from triage @joshtriplett any updates?

**Comment by Lucretiel** [comment]
Sorry for the delay, I didn't realize my fork wasn't updated with upstream master, and so I believed it when it said I was all up to date after a pull + rebase. There was a pretty substantial refactor of the `core::iter` file layout so I'll need to migrate to that; I'll try to take care of it tonight.

**Comment by Dylan-DPC** [comment]
ping from triage @Lucretiel any updates?

**Comment by jonas-schievink** [comment]
Closing due to inactivity, but feel free to resubmit with the necessary changes!

**Comment by Lucretiel** [comment]
Done! Will re-opening the PR automatically bring in my latest branch changes?

**Comment by jonas-schievink** [comment]
@Lucretiel I think you'll have to open another PR

**Pull request closed without merging**